datashader/transfer_functions/__init__.py

```diff
@@ -391,50 +393,71 @@ def _colorize(agg, color_key, how, alpha, span, min_alpha, name, color_baseline,
     data = agg_t.data.transpose([1, 2, 0])
     if da and isinstance(data, da.Array):
         data = data.compute()
-    color_data = data.copy()
+
+    H, W, C = data.shape
+    color_data = ascontigous(data.copy())
+
+    # Keep a NaN mask for later, and a non-NaN mask for averages
+    nan_mask = np.isnan(data)
+    color_mask = ~nan_mask  # bool (H, W, C)
 
     # subtract color_baseline if needed
     with warnings.catch_warnings():
         warnings.filterwarnings('ignore', r'All-NaN slice encountered')
         baseline = np.nanmin(color_data) if color_baseline is None else color_baseline
     with np.errstate(invalid='ignore'):
+        # in-place add/sub to minimize temporaries
         if baseline > 0:
-            color_data -= baseline
+            np.subtract(color_data, baseline, out=color_data, where=color_mask)
         elif baseline < 0:
-            color_data += -baseline
-        if color_data.dtype.kind != 'u' and color_baseline is not None:
-            color_data[color_data<0]=0
+            np.add(color_data, -baseline, out=color_data, where=color_mask)
 
-    color_total = nansum_missing(color_data, axis=2)
-    # dot does not handle nans, so replace with zeros
-    color_data[np.isnan(data)] = 0
+    # If an explicit baseline was given and dtype is signed, clip negatives to 0 (in-place)
+    if (color_baseline is not None) and (color_data.dtype.kind != 'u'):
+        np.maximum(color_data, 0, out=color_data)
 
-    # zero-count pixels will be 0/0, but it's safe to ignore that when dividing
+    # --- Prepare totals and zero-out NaNs once ---
+    # Use sum along last axis; NaNs contribute 0 after zeroing
+    color_total = np.nansum(color_data, axis=2)
+
+    # Replace NaNs with 0s for dot/matmul in one pass (in-place)
+    # If you don't want to mutate color_data contents further, copy first.
+    np.nan_to_num(color_data, copy=False)  # NaN -> 0
```

Comment:
Can't we use an existing mask for this? I would assume `nan_to_mask` would do this calculation behind the scenes. 

Also note that this will convert `inf` values to floats.